File: configuration/src/main/java/com/synopsys/integration/configuration/property/Property.java
package com.synopsys.integration.configuration.property;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.synopsys.integration.configuration.property.deprecation.DeprecatedValueUsage;
import com.synopsys.integration.configuration.property.deprecation.PropertyDeprecationInfo;
import com.synopsys.integration.configuration.property.deprecation.PropertyRemovalDeprecationInfo;
import com.synopsys.integration.configuration.util.Category;
import com.synopsys.integration.configuration.util.Group;
import com.synopsys.integration.configuration.util.ProductMajorVersion;

/**
 * This is the most basic property.
 * It has no type information and a value cannot be retrieved for it (without a subclass).
 **/
public abstract class Property {
    public Property(String key) {
        this.key = key;
    }

    @NotNull
    private final String key;
    @Nullable
    private String name = null;
    @Nullable
    private String fromVersion = null;
    @Nullable
    private PropertyHelpInfo propertyHelpInfo = null;
    @Nullable
    private PropertyGroupInfo propertyGroupInfo = null;
    @Nullable
    private Category category = null;
    @NotNull
    private final PropertyDeprecationInfo propertyDeprecationInfo = new PropertyDeprecationInfo();
    @Nullable
    private String example = null;

    public Property setInfo(String name, String fromVersion) {
        this.name = name;
        this.fromVersion = fromVersion;
        return this;
    }

    public Property setHelp(@NotNull String shortText) {
        return setHelp(new PropertyHelpInfo(shortText, null));
    }

    public Property setHelp(@NotNull String shortText, @Nullable String longText) {
        return setHelp(new PropertyHelpInfo(shortText, longText));
    }

    public Property setHelp(PropertyHelpInfo propertyHelpInfo) {
        this.propertyHelpInfo = propertyHelpInfo;
        return this;
    }

    public Property setGroups(Group primaryGroup, Group... additionalGroups) {
        return setGroups(new PropertyGroupInfo(primaryGroup, additionalGroups));
    }

    public Property setGroups(PropertyGroupInfo propertyGroupInfo) {
        this.propertyGroupInfo = propertyGroupInfo;
        return this;
    }

    public Property setCategory(Category category) {
        this.category = category;
        return this;
    }

    public Property setRemovalDeprecation(String description, ProductMajorVersion removeInVersion) {
        propertyDeprecationInfo.setRemovalDeprecation(description, removeInVersion);
        return this;
    }

    public void addDeprecatedValueInfo(String valueDescription, String reason) {
        propertyDeprecationInfo.addDeprecatedValueInfo(valueDescription, reason);
    }

    protected Optional<DeprecatedValueUsage> createDeprecatedValueUsageIfExists(String valueDescription) {
        return getPropertyDeprecationInfo().getDeprecatedValues().stream()
            .filter(info -> info.getValueDescription().equals(valueDescription))
            .findFirst()
            .map(info -> new DeprecatedValueUsage(valueDescription, info));
    }

    public Property setExample(String example) {
        this.example = example;
        return this;
    }

    public boolean isCaseSensitive() {
        return false;
    }

    public boolean isOnlyExampleValues() {
        return false;
    }

    public List<String> listExampleValues() {
        return new ArrayList<>();
    }

    public String describeType() {
        return null;
    }

    public String describeDefault() {
        return null;
    }

    public boolean isCommaSeparated() {
        return false;
    }

    public String getKey() {
        return key;
    }

    public String getName() {
        return name;
    }

    public String getFromVersion() {
        return fromVersion;
    }

    public PropertyHelpInfo getPropertyHelpInfo() {
        return propertyHelpInfo;
    }

    public PropertyGroupInfo getPropertyGroupInfo() {
        return propertyGroupInfo;
    }

    public Category getCategory() {
        return category;
    }

    public PropertyDeprecationInfo getPropertyDeprecationInfo() {
        return propertyDeprecationInfo;
    }

    public boolean isDeprecatedForRemoval() {
        return propertyDeprecationInfo.getRemovalInfo().isPresent();
    }

    public String getKeyAsEnvironmentVariable() {
        if (StringUtils.isNotBlank(key)) {
            return key.replace(".", "_").toUpperCase();
        }

        return key;
    }

    public String getExample() {
        return example;
    }

    public void setRemovalDeprecation(PropertyRemovalDeprecationInfo propertyRemovalDeprecationInfo) {
        this.propertyDeprecationInfo.setRemovalDeprecation(propertyRemovalDeprecationInfo);
    }
}

File: configuration/src/main/java/com/synopsys/integration/configuration/property/Property.kt
/**
 * configuration
 *
 * Copyright (c) 2020 Synopsys, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package com.synopsys.integration.configuration.property

/**
 * This is the most basic property.
 * It has no type information and a value cannot be retrieved for it (without a subclass).
 **/
abstract class Property(val key: String) {
    var name: String? = null
    var fromVersion: String? = null
    var propertyHelpInfo: PropertyHelpInfo? = null
    var propertyGroupInfo: PropertyGroupInfo? = null
    var category: Category? = null
    var propertyDeprecationInfo: PropertyDeprecationInfo? = null

    fun info(name: String, fromVersion: String): Property {
        this.name = name
        this.fromVersion = fromVersion
        return this
    }

    fun help(short: String, long: String? = null): Property {
        this.propertyHelpInfo = PropertyHelpInfo(short, long)
        return this
    }

    fun groups(primaryGroup: Group, vararg additionalGroups: Group): Property {
        this.propertyGroupInfo = PropertyGroupInfo(primaryGroup, additionalGroups.toList())
        return this
    }

    fun category(category: Category): Property {
        this.category = category
        return this
    }

    fun deprecated(description: String, failInVersion: ProductMajorVersion, removeInVersion: ProductMajorVersion): Property {
        this.propertyDeprecationInfo = PropertyDeprecationInfo(description, failInVersion, removeInVersion)
        return this
    }

    open fun isCaseSensitive(): Boolean = false
    open fun isOnlyExampleValues(): Boolean = false
    open fun listExampleValues(): List<String>? = emptyList()
    open fun describeType(): String? = null
    open fun describeDefault(): String? = null
    open fun isCommaSeparated(): Boolean = false
}

data class PropertyHelpInfo(val short: String, val long: String?)
data class PropertyGroupInfo(val primaryGroup: Group, val additionalGroups: List<Group>)
data class PropertyDeprecationInfo(val description: String, val failInVersion: ProductMajorVersion, val removeInVersion: ProductMajorVersion) {
    fun getDeprecationText(): String = "$description It will cause failure in ${failInVersion.getDisplayValue()} and be removed in ${removeInVersion.getDisplayValue()}.";
}

